### PR TITLE
#15: refactor write pipeline

### PR DESCRIPTION
**Pull request by dpc**
This prepares `rdedup` to pararelize just as described in #15

### CODE REVIEW

**Comment by dpc** [comment]
I just got it to work. It hasn't been actually tested very much yet but I guess it wouldn't hurt to get it reviewed. As usually in Rust, after all the work to actually figure it out how to get it to compile, it passes all the new and existing tests.

It actually removes the parallelism (TODO included) http://stackoverflow.com/questions/42476389/how-to-parallely-map-on-a-custom-single-threaded-iterator-in-rust .

I've introduced new, hopefully efficient way, to chunk the buffers without ever having to copy any data needlessly. Everything operates now on scatter-gather buffers - reference counted (`Arc`) views into parts of buffers read from the input.

The `sg.rs` module is clumsy, and was introduced so I can work on new pieces without changing the existing ones. I think all stuff there will need to be more logically structured.

I'm afraid of bugs, but let's be hopeful that type system will protect us. :)

I'll followup when I figure out how to actually run it in parallel. Then we can start measuring.

**Comment by phillipCouto** [comment]
@dpc excellent! I took a peek and love the iterator approach. The iterators will allow for an elegant business logic composer in the main lib.rs  making it easier to read.

I will go through it in more detail tomorrow.

**Comment by dpc** [comment]
SMP stuff added. I have spent some time investigated the performance and biggest bottleneck was IO, so it's now working in a pool. There is still a lot of room for improvement, but on my system the difference is like an order of magnitude already.

File: lib/src/sg.rs [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+#[derive(Debug, PartialEq, Eq)]
+pub struct Chunk {
+    sg: SGBuf,
+    chunk_type: DataType,
+    data_type: DataType,
+}
+
+/// Scattered-gathered buffer
+///
+/// A pice of data potentially scattered between
```
Comment (phillipCouto) [inline lib/src/sg.rs]: Change ```pice``` to ```piece```

File: lib/src/sg.rs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+use std::ops::{Deref, DerefMut};
+use std::result::Result;
+use std::sync::Arc;
+
+pub trait EdgeFinder {
+    fn find_edges(&mut self, buf: &[u8]) -> Vec<usize>;
+}
+
+/// Finds edges using rolling sum
+pub struct BupEdgeFinder {
```
Comment (phillipCouto) [inline lib/src/sg.rs]: NVM I saw the edge finder at the end of the file.

**Comment by phillipCouto** [comment]
We turned stable back on for the builds?

File: lib/src/lib.rs [hunk 3]
```diff
@@ -1134,194 +1092,111 @@ impl Repo {
...
-                        .unwrap();
-                }
+        crossbeam::scope(|scope| {
+            // Unlike CPU-intense workload, it's hard to come up with one
+            // formula for best number of `fdatasync` threads for everyone.
+            // Each buffer is blocked on one chunk, so the longer the chunks,
+            // the less threads are needed to saturate the disk and so on. Some
+            // form of run-time scaling would be best. I picked some reasonable
+            // that also works well for me with defaults. --dpc
+            let thread_num = 4 * self.write_cpu_thread_num();
```
Comment (dpc) [inline lib/src/lib.rs]: A subject for future investigation. I 'd like to add logging-based performance reporting first so it's easier to figure out what works, and what not.

**Comment by dpc** [comment]
Yes, I turned on `stable` and `beta` again without putting it into separate PR. Now I see it actualy fails because of clippy? Hmm...

**Comment by phillipCouto** [comment]
Ah ok, clippy only works in nightly. We need to update the ci config not to run clippy when in beta or stable.

**Comment by phillipCouto** [comment]
Besides the comment spelling error it is good to go =).

**Pull request merged**